pkg/gce-pd-csi-driver/controller.go

```diff
@@ -549,6 +709,83 @@ func (gceCS *GCEControllerServer) DeleteVolume(ctx context.Context, req *csi.Del
 		return &csi.DeleteVolumeResponse{}, nil
 	}
 
+	volumeIsMultiZone := isMultiZoneVolKey(volKey)
+	if gceCS.multiZoneVolumeHandleConfig.Enable && volumeIsMultiZone {
+		// Delete multi-zone disk, that may have up to N disks.
+		return gceCS.deleteMultiZoneDisk(ctx, req, project, volKey)
+	}
+
+	// Delete zonal or regional disk
+	return gceCS.deleteSingleDeviceDisk(ctx, req, project, volKey)
+}
+
+func getGCEApiVersion(multiWriter bool) gce.GCEAPIVersion {
+	if multiWriter {
+		return gce.GCEAPIVersionBeta
+	}
+
+	return gce.GCEAPIVersionV1
+}
+
+func (gceCS *GCEControllerServer) deleteMultiZoneDisk(ctx context.Context, req *csi.DeleteVolumeRequest, project string, volKey *meta.Key) (*csi.DeleteVolumeResponse, error) {
+	// List disks with same name
+	var err error
+	diskTypeForMetric := metrics.DefaultDiskTypeForMetric
+	enableConfidentialCompute := metrics.DefaultEnableConfidentialCompute
+	enableStoragePools := metrics.DefaultEnableStoragePools
+	defer func() {
+		gceCS.Metrics.RecordOperationErrorMetrics("DeleteVolume", err, diskTypeForMetric, enableConfidentialCompute, enableStoragePools)
+	}()
+	existingZones := []string{gceCS.CloudProvider.GetDefaultZone()}
+	zones, err := getDefaultZonesInRegion(ctx, gceCS, existingZones)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list default zones: %w", err)
+	}
+
+	volumeID := req.GetVolumeId()
+	if acquired := gceCS.volumeLocks.TryAcquire(volumeID); !acquired {
+		return nil, status.Errorf(codes.Aborted, common.VolumeOperationAlreadyExistsFmt, volumeID)
+	}
+	defer gceCS.volumeLocks.Release(volumeID)
+
+	deleteDiskErrs := []error{}
+	for _, zone := range zones {
+		zonalVolKey := &meta.Key{
+			Name:   volKey.Name,
+			Region: volKey.Region,
+			Zone:   zone,
+		}
+		disk, _ := gceCS.CloudProvider.GetDisk(ctx, project, zonalVolKey, gce.GCEAPIVersionV1)
+		// TODO: Consolidate the parameters here, rather than taking the last.
+		diskTypeForMetric, enableConfidentialCompute, enableStoragePools = metrics.GetMetricParameters(disk)
+		err := gceCS.CloudProvider.DeleteDisk(ctx, project, zonalVolKey)
+		if err != nil {
+			deleteDiskErrs = append(deleteDiskErrs, gceCS.CloudProvider.DeleteDisk(ctx, project, volKey))
+		}
+	}
+
+	if len(deleteDiskErrs) == 1 {
+		return nil, common.LoggedError("Failed to delete disk: ", err)
+	} else if len(deleteDiskErrs) > 1 {
+		// Return a non-final error here, since one of the errors could be a non-final.
```

Comment:
In contrast, endlessly trying to delete disks is probably desired behavior.

Reply:
Updated to check if the error is one of the retryable codes.